internal/dataplane/parser/translate_grpcroute.go

```diff
@@ -0,0 +1,133 @@
+	// first we grab the spec and gather some metdata about the object
+	spec := grpcroute.Spec
+
+	if len(spec.Rules) == 0 {
+		return errRouteValidationNoRules
+	}
+
+	// each rule may represent a different set of backend services that will be accepting
+	// traffic, so we make separate routes and Kong services for every present rule.
+	for ruleNumber, rule := range spec.Rules {
+		// determine the routes needed to route traffic to services for this rule
+		routes := generateKongRoutesFromGRPCRouteRule(grpcroute, ruleNumber, rule)
+
+		// create a service and attach the routes to it
+		service, err := generateKongServiceFromBackendRefWithRuleNumber(p.logger, p.storer, result, grpcroute, ruleNumber, "grpcs", grpcBackendRefsToBackendRefs(rule.BackendRefs)...)
+		if err != nil {
+			return err
+		}
+		service.Routes = append(service.Routes, routes...)
+
+		// cache the service to avoid duplicates in further loop iterations
+		result.ServiceNameToServices[*service.Service.Name] = service
+	}
+
+	return nil
+}
+
+func generateKongRoutesFromGRPCRouteRule(grpcroute *gatewayv1alpha2.GRPCRoute, ruleNumber int, rule gatewayv1alpha2.GRPCRouteRule) []kongstate.Route {
+	var routes = make([]kongstate.Route, 0, len(rule.Matches))
+
+	// gather the k8s object information and hostnames from the httproute
+	ingressObjectInfo := util.FromK8sObject(grpcroute)
+
+	for matchNumber := range rule.Matches {
+		routeName := fmt.Sprintf(
+			"httproute.%s.%s.%d.%d",
+			grpcroute.Namespace,
+			grpcroute.Name,
+			ruleNumber,
+			matchNumber,
+		)
+
+		r := kongstate.Route{
+			Ingress: ingressObjectInfo,
+			Route: kong.Route{
+				Name:      kong.String(routeName),
+				Protocols: kong.StringSlice("grpc", "grpcs"),
+			},
+		}
+
+		if len(grpcroute.Spec.Hostnames) > 0 {
+			r.Hosts = getGRPCRouteHostnamesAsSliceOfStringPointers(grpcroute)
+		}
+
+		routes = append(routes, r)
+	}
+
+	return routes
+}
+
+func grpcBackendRefsToBackendRefs(grpcBackendRef []gatewayv1alpha2.GRPCBackendRef) []gatewayv1beta1.BackendRef {
+	backendRefs := make([]gatewayv1beta1.BackendRef, 0, len(grpcBackendRef))
+
+	for _, hRef := range grpcBackendRef {
+		backendRefs = append(backendRefs, hRef.BackendRef)
+	}
+	return backendRefs
+}
+
+// -----------------------------------------------------------------------------
+// Translate HTTPRoute - Utils
+// -----------------------------------------------------------------------------
+
+// getHTTPRouteHostnamesAsSliceOfStringPointers translates the hostnames defined
+// in an HTTPRoute specification into a []*string slice, which is the type required
+// by kong.Route{}.
+func getGRPCRouteHostnamesAsSliceOfStringPointers(httproute *gatewayv1alpha2.GRPCRoute) []*string {
+	hostnames := make([]*string, 0, len(httproute.Spec.Hostnames))
+	for _, hostname := range httproute.Spec.Hostnames {
+		hostnames = append(hostnames, kong.String(string(hostname)))
+	}
+	return hostnames
```

Comment:
As a happy `github.com/samber/lo` library user 🙂, I may suggest using `lo.Map` in such cases. In this particular situation that would be:

```
return lo.Map(httproute.Spec.Hostnames, func(h gatewayv1beta1.Hostname, _ int) *string {
	return lo.ToPtr(string(h))
})
```

No hard feelings if you drop this suggestion though. 👍

Reply:
that looks nice, I'll give it a try